aten/src/ATen/Storage.h

```diff
@@ -7,7 +7,7 @@ namespace at {
 struct AT_API Storage {
 public:
   Storage() = delete;
-  Storage(StorageImpl* storage_impl) : storage_impl_(storage_impl) {}
+  Storage(StorageImpl* storage_impl) : storage_impl_(c10::intrusive_ptr<StorageImpl>::reclaim(storage_impl)) {}
```

Comment:
so, the documentation of reclaim says "the pointer passed in *must* have been created using intrusive_ptr::release()" -- how do you know this is true?

Reply:
You don't: it's an invariant calling code has to uphold. Essentially, we have both raw non-owning pointers and raw "owning" pointers, and it's only valid to call reclaim on raw owning pointers; calling it on a non-owning pointer will result in the refcount being less than the true refcount. There is no dynamic check for this (until you double free...)

There is also some business with stack allocated objects.

There are two ways a `T*` (where T inherits from `intrusive_ptr_target`) can be constructed. First is the ordinary way: you `new T` or stack allocate it. In this case, the reference count of T (in `intrusive_ptr_target` is initialized to zero as a special "sentinel" value; it doesn't mean that the refcount is actually zero, just that the object isn't managed by `intrusive_ptr`. `reclaim()` checks if the refcount is zero, and fails an assert if it is.

Second is via `make_intrusive`; in this case, the refcount is set to one, and the object will be destructed when its refcount goes to zero. In this case, if you get a raw pointer from it using `release()`, you can turn it back into an intrusive pointer with reclaim, and the refcount will be non-zero. (The danger case is when you get a non-owning raw pointer out using `get()`)